Compilation error on 32-bit system

ESP32, GCC 12.2 (32-bit target)

Compiling code that serializes structures with arrays/vectors results in errors like that:

```cpp
glaze-src/include/glaze/util/dump.hpp:41:32: error: no matching function for call to 'max(std::__cxx11::basic_string<char>::size_type, const long unsigned int&)'
   41 |             b.resize((std::max)(b.size() * 2, k));
      |                      ~~~~~~~~~~^~~~~~~~~~~~~~~~~
In file included from include/c++/12.2.0/string:50,
                 from include/c++/12.2.0/bits/locale_classes.h:40,
                 from include/c++/12.2.0/bits/ios_base.h:41,
                 from include/c++/12.2.0/ios:42,
                 from include/c++/12.2.0/istream:38,
                 from include/c++/12.2.0/sstream:38,
                 from include/c++/12.2.0/chrono:41,
                 from base/traits/types.hpp:5,
                 from base/meta/runtime.hpp:3:
include/c++/12.2.0/bits/stl_algobase.h:254:5: note: candidate: 'template<class _Tp> constexpr const _Tp& std::max(const _Tp&, const _Tp&)'
  254 |     max(const _Tp& __a, const _Tp& __b)
      |     ^~~
include/c++/12.2.0/bits/stl_algobase.h:254:5: note:   template argument deduction/substitution failed:
glaze-src/include/glaze/util/dump.hpp:41:32: note:   deduced conflicting types for parameter 'const _Tp' ('unsigned int' and 'long unsigned int')
   41 |             b.resize((std::max)(b.size() * 2, k));
      |                      ~~~~~~~~~~^~~~~~~~~~~~~~~~~
include/c++/12.2.0/bits/stl_algobase.h:300:5: note: candidate: 'template<class _Tp, class _Compare> constexpr const _Tp& std::max(const _Tp&, const _Tp&, _Compare)'
  300 |     max(const _Tp& __a, const _Tp& __b, _Compare __comp)
      |     ^~~
include/c++/12.2.0/bits/stl_algobase.h:300:5: note:   template argument deduction/substitution failed:
glaze-src/include/glaze/util/dump.hpp:41:32: note:   deduced conflicting types for parameter 'const _Tp' ('unsigned int' and 'long unsigned int')
   41 |             b.resize((std::max)(b.size() * 2, k));
      |                      ~~~~~~~~~~^~~~~~~~~~~~~~~~~
In file included from include/c++/12.2.0/functional:64,
                 from glaze-src/include/glaze/util/type_traits.hpp:6,
                 from glaze-src/include/glaze/core/opts.hpp:8,
                 from glaze-src/include/glaze/binary/skip.hpp:7:
include/c++/12.2.0/bits/stl_algo.h:5746:5: note: candidate: 'template<class _Tp> constexpr _Tp std::max(initializer_list<_Tp>)'
 5746 |     max(initializer_list<_Tp> __l)
      |     ^~~
include/c++/12.2.0/bits/stl_algo.h:5746:5: note:   template argument deduction/substitution failed:
glaze-src/include/glaze/util/dump.hpp:41:32: note:   mismatched types 'std::initializer_list<_Tp>' and 'unsigned int'
   41 |             b.resize((std::max)(b.size() * 2, k));
      |                      ~~~~~~~~~~^~~~~~~~~~~~~~~~~
include/c++/12.2.0/bits/stl_algo.h:5756:5: note: candidate: 'template<class _Tp, class _Compare> constexpr _Tp std::max(initializer_list<_Tp>, _Compare)'
 5756 |     max(initializer_list<_Tp> __l, _Compare __comp)
      |     ^~~
include/c++/12.2.0/bits/stl_algo.h:5756:5: note:   template argument deduction/substitution failed:
glaze-src/include/glaze/util/dump.hpp:41:32: note:   mismatched types 'std::initializer_list<_Tp>' and 'unsigned int'
   41 |             b.resize((std::max)(b.size() * 2, k));
```

Casting to std::size_t resolves issue as `std::max` expects both values to be of the same type:
```cpp
   template <uint32_t N, class B>
   GLZ_ALWAYS_INLINE void maybe_pad(B& b, auto& ix) noexcept
   {
      if constexpr (vector_like<B>) {
         if (const auto k = ix + N; k > b.size()) [[unlikely]] {
            b.resize((std::max)(std::size_t(b.size() * 2), std::size_t(k))); // cast values to std::size_t before passing to std::max
         }
      }
   }
   ```
   
While this may work for most use cases, it indicates deeper issue with glaze + 32-bit - most likely some fixes are needed to use std::size_t instead, as this would work nicely for both 32-bit and 64-bit systems?

--- a/.github/workflows/gcc-x86.yml
+++ b/.github/workflows/gcc-x86.yml
@@ -1,0 +1,45 @@
+name: gcc-x86
+
+on:
+   push:
+      branches:
+         - main
+         - feature/*
+      paths-ignore:
+         - "**.md"
+   pull_request:
+      branches: [main]
+      paths-ignore:
+         - "**.md"
+   workflow_dispatch:
+
+jobs:
+   build:
+      runs-on: ubuntu-24.04
+
+      strategy:
+         fail-fast: false
+         matrix:
+            gcc: [12, 13, 14]
+            build_type: [Debug]
+            std: [23]
+
+      env:
+         CC: gcc-${{matrix.gcc}}
+         CXX: g++-${{matrix.gcc}}
+
+      steps:
+         - uses: actions/checkout@v4
+
+         - name: Setup environment
+           run: |
+              sudo apt update
+              sudo apt-get install -y gcc-${{matrix.gcc}}-multilib g++-${{matrix.gcc}}-multilib libc6-dev-i386
+         - name: Configure CMake
+           run: |
+              CXXFLAGS="-g3 -m32" cmake -B ${{github.workspace}}/build -DCMAKE_BUILD_TYPE=${{matrix.build_type}} -DCMAKE_CXX_STANDARD=${{matrix.std}}
+         - name: Build
+           run: cmake --build build -j $(nproc)
+
+         - name: Test
+           run: ctest --output-on-failure --test-dir ${{github.workspace}}/build

--- a/include/glaze/binary/header.hpp
+++ b/include/glaze/binary/header.hpp
@@ -59,7 +59,7 @@ namespace glz::detail
 
    inline constexpr std::array<uint8_t, 8> byte_count_lookup{1, 2, 4, 8, 16, 32, 64, 128};
 
-   [[nodiscard]] GLZ_ALWAYS_INLINE constexpr uint64_t int_from_compressed(auto&& ctx, auto&& it, auto&& end) noexcept
+   [[nodiscard]] GLZ_ALWAYS_INLINE constexpr size_t int_from_compressed(auto&& ctx, auto&& it, auto&& end) noexcept
    {
       if (it >= end) [[unlikely]] {
          ctx.error = error_code::unexpected_end;
@@ -92,16 +92,22 @@ namespace glz::detail
          return h >> 2;
       }
       case 3: {
-         uint64_t h;
-         std::memcpy(&h, it, 8);
-         it += 8;
-         h = h >> 2;
-         static constexpr uint64_t safety_limit = 1ull << 48; // 2^48
-         if (h > safety_limit) [[unlikely]] {
-            ctx.error = error_code::unexpected_end;
-            return 0;
+         // On 32-bit systems it's impossible to address more than 4 GBiB of memory, so we should verify first if we are
+         // running in 64-bit mode here
+         if constexpr (sizeof(size_t) > sizeof(uint32_t)) {
+            uint64_t h;
+            std::memcpy(&h, it, 8);
+            it += 8;
+            h = h >> 2;
+            static constexpr uint64_t safety_limit = 1ull << 48; // 2^48
+            if (h > safety_limit) [[unlikely]] {
+               ctx.error = error_code::unexpected_end;
+               return 0;
+            }
+            return h;
          }
-         return h;
+         // Fallthrough in case we are in 32-bit mode
+         [[fallthrough]];
       }
       default:
          return 0;

--- a/include/glaze/core/refl.hpp
+++ b/include/glaze/core/refl.hpp
@@ -1077,8 +1077,8 @@ namespace glz::detail
    }
 
    // Do not call this at runtime, it is assumes the key lies within min_length and max_length
-   inline constexpr size_t full_hash_impl(const sv key, const uint64_t seed, const auto min_length,
-                                          const auto max_length) noexcept
+   inline constexpr uint64_t full_hash_impl(const sv key, const uint64_t seed, const auto min_length,
+                                            const auto max_length) noexcept
    {
       if (max_length < 8) {
          return bitmix(to_uint64_n_below_8(key.data(), key.size()), seed);
@@ -1113,8 +1113,8 @@ namespace glz::detail
    }
 
    // runtime full hash algorithm
-   template <size_t min_length, size_t max_length, size_t seed>
-   inline constexpr size_t full_hash(const auto* it, const size_t n) noexcept
+   template <uint64_t min_length, uint64_t max_length, uint64_t seed>
+   inline constexpr uint64_t full_hash(const auto* it, const size_t n) noexcept
    {
       if constexpr (max_length < 8) {
          if (n > 7) {

--- a/tests/api_test/api_test.cpp
+++ b/tests/api_test/api_test.cpp
@@ -204,7 +204,12 @@ void tests()
 
    "span type name"_test = [] {
       std::string_view s = glz::name_v<std::span<double>>;
-      expect(s == "std::span<double,18446744073709551615>");
+      if constexpr (sizeof(size_t) == sizeof(uint64_t)) {
+         expect(s == "std::span<double,18446744073709551615>");
+      }
+      else if constexpr (sizeof(size_t) == sizeof(uint32_t)) {
+         expect(s == "std::span<double,4294967295>");
+      }
    };
 
    "tuple type name"_test = [] {

--- a/tests/lib_test/lib_test.cpp
+++ b/tests/lib_test/lib_test.cpp
@@ -81,7 +81,12 @@ void tests()
 
    "span type name"_test = [] {
       std::string_view s = glz::name_v<std::span<double>>;
-      expect(s == "std::span<double,18446744073709551615>");
+      if constexpr (sizeof(size_t) == sizeof(uint64_t)) {
+         expect(s == "std::span<double,18446744073709551615>");
+      }
+      else if constexpr (sizeof(size_t) == sizeof(uint32_t)) {
+         expect(s == "std::span<double,4294967295>");
+      }
    };
 
    "my_api type io"_test = [&] {

